FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean

```diff
@@ -785,9 +779,210 @@ theorem adicCompletionComapAlgEquiv_integral :
+
+  have : LinearMap.CompatibleSMul (L ⊗[K] adicCompletion K v)
+      ((w : Extension B v) → adicCompletion L w.val)
+      (adicCompletionIntegers K v) (adicCompletion K v) := {
+    map_smul f x y := by
+      obtain ⟨x, hx⟩ := x
+      exact LinearMap.map_smul f x y
+  }
+  let f : L ⊗[K] adicCompletion K v →ₗ[adicCompletionIntegers K v]
+      (w' : Extension B v) → adicCompletion L w'.val :=
+    (tensorAdicCompletionComapLinearMap A K L B v).restrictScalars (adicCompletionIntegers K v)
+  let g := LinearMap.comp f (h A K L B v)
+  @LinearMap.proj _ _ _ (fun (w : Extension B v) ↦ adicCompletion L w.val) _
+    (fun w => (alg w).toModule) ⟨w, hvw⟩ |>.comp g
+
+attribute [local instance] Algebra.TensorProduct.rightAlgebra in
+noncomputable def adicCompletionComapIntLinearMap_range_eq_integers (v : HeightOneSpectrum A)
+    (w : HeightOneSpectrum B) (hvw : comap A w = v) :
+    letI := comap_algebra A K L B hvw
+    Set.range (adicCompletionComapIntLinearMap A K L B v w hvw) = adicCompletionIntegers L w := by
+  ext x
+  have memrange : (Set.range _) = _ :=
+    congr_arg SetLike.coe (adicCompletionComapAlgEquiv_integral A K L B v)
+  rw [Subalgebra.coe_pi, Submodule.carrier_eq_coe, Submodule.coe_pi, Set.ext_iff] at memrange
+  let extension : Extension B v := ⟨w, hvw⟩
+  constructor
+  . rintro ⟨y, rfl⟩
+    exact (memrange _).mp (Set.mem_range_self y) extension trivial
+  . intro hx
+    classical
+    let x' : (w : Extension B v) → adicCompletion L w.val := Pi.single extension x
+    obtain ⟨y, (hy : _ = x')⟩ : x' ∈ Set.range _ := by
+      rw [memrange x', Set.mem_pi]
+      intro w _
+      dsimp [x']
+      by_cases h : extension = w
+      . rw [← h]
+        simpa
+      . rw [Pi.single_eq_of_ne' h]
+        apply Subalgebra.zero_mem
+    use y
+    simpa [adicCompletionComapIntLinearMap, x', extension] using congr_fun hy ⟨w, hvw⟩
+
+attribute [local instance] Algebra.TensorProduct.rightAlgebra in
+/-- `𝓞_w` is finite over `𝓞_v`. -/
+-- This can be proved for finite extensions of complete discretely valued fields without
+-- referencing the underlying fields being completed, but this is sufficient for our
+-- purposes.
+theorem comap_int_algebra_finite (v : HeightOneSpectrum A) (w : HeightOneSpectrum B)
+    (hvw : comap A w = v) :
+    letI := comap_int_algebra A K L B hvw
+    Module.Finite (adicCompletionIntegers K v) (adicCompletionIntegers L w) := by
+  letI := comap_int_algebra A K L B hvw
+  letI := comap_algebra A K L B hvw
+  letI := RingHom.toAlgebra <| (algebraMap (adicCompletion K v) (adicCompletion L w)).comp
+      (algebraMap (adicCompletionIntegers K v) (adicCompletion K v))
+  have : IsScalarTower (adicCompletionIntegers K v) (adicCompletion K v)
+      (adicCompletion L w) := by
+    apply IsScalarTower.of_algebraMap_smul fun _ _ ↦ rfl
+
+  let f := adicCompletionComapIntLinearMap A K L B v w hvw
+  let integerSubmodule : Submodule (adicCompletionIntegers K v) (adicCompletion L w) := {
```

Comment:
I couldn't find anything helpful here in Mathlib

Reply:
I've moved this to `FLT/Mathlib/Algebra/Module/Submodule/Basic.lean`, but still can't find it in Mathlib (note: I want the carrier to be def-eq to the subring because it makes things easier to prove, so something defined in terms of ranges isn't helpful)